### PR TITLE
refactor(app): protocol Upload Empty State Recent Run Listing

### LINKED ISSUE #8464: Protocol Upload Empty State Recent Run Listing
## Overview

The Protocol Upload page's empty state needs a couple of additions that direct users to new rerun protocol CTA.

empty state:
https://www.figma.com/file/JCQNOTTK9tTmYPhRqeOfSj/Milestone-0:-Protocol-Upload-Revamp?node-id=3205%3A22

empty state if a user has not run a protocol on this robot before:
https://www.figma.com/file/JCQNOTTK9tTmYPhRqeOfSj/Milestone-0:-Protocol-Upload-Revamp?node-id=6957%3A167822

## Acceptance Criteria

- [ ] Title bar contents should be "Open a protocol to run on {robotName}"
- [ ] Only if a protocol has been run on this robot before add a section between the file drop zone and the "Don't have a protocol?" link separated by dividers. The contents of the section should be as follows: 
   - [ ] Add header for "{robotName}'s last run". 
   - [ ] Add table with one row of data and the following columns:
     - [ ] "Protocol name" (protocol's name)
     - [ ] "Run timestamp" (timestamp of the last run with TZ)
     - [ ] If any offset data exists "Labware Offset data" ("{numOffsets} Labware Offsets"), otherwise omit this column and value.
   - [ ] The final column should be unlabeled and filled with a CTA to rerun this protocol labeled "Run Again". 
   - [ ] add "See How Rerunning a Protocol Works" link which launches an explanatory closable modal with a single "Close" CTA at the bottom. The modal should have an external link out to "Learn More about Labware Offset Data" that directs to an explanatory support article.  
- [ ] The "Don't have a protocol yet?" section should be updated as follows:
   - [ ] "Launch Protocol Library" with link out icon 
   - [ ] "Launch Protocol Designer" with link out icon

### PR DESCRIPTION
**Pull request by jerader**
# Overview

closes #8464 

# Changelog

- creates  `RerunningProtocolModal` and adds on to `UploadInfo` component 

# Review requests

- review AC on ticket and review styling to make sure it matches [figma](https://www.figma.com/file/JCQNOTTK9tTmYPhRqeOfSj/Milestone-0%3A-Protocol-Upload-Revamp?node-id=6957%3A167940)
- the link in the `RerunningProtocolModal` is empty for now since we are still waiting on support to write the article

# Risk assessment

low, behind ff

### CODE REVIEW

**Comment by jerader** [comment]
@emilywools - here is the modal + the empty state with and without previous run data. 

Screen Shot 2021-11-12 at 16 33 04
Screen Shot 2021-11-12 at 16 20 17
Screen Shot 2021-11-12 at 16 33 24

**Comment by emilywools** [comment]
Hey @jerader this is looking great. can you add this line as the 3rd paragraph in the modal window (after the second paragraph)?

> If you recalibrate your robot, it will clear the last run from the upload page.

**Comment by emilywools** [comment]
@jerader on the "launch protocol library" link, Library needs to be capitalized

on the "see how x works" it needs to be title case

File: app/src/organisms/ProtocolUpload/UploadInput.tsx [hunk 1]
```diff
@@ -101,23 +134,38 @@ export function UploadInput(props: UploadInputProps): JSX.Element {
...
+  const fullRunTimestamp = runQuery.data?.data.createdAt
+  const indexToSplit = fullRunTimestamp?.indexOf('T')
+  if (fullRunTimestamp == null) return null
+  const date = fullRunTimestamp?.slice(0, indexToSplit)
+  const timeAndUTC =
+    indexToSplit != null ? fullRunTimestamp?.slice(indexToSplit + 1) : ''
+  const timeToSplit = timeAndUTC.indexOf('.')
+  const time = timeAndUTC.slice(0, timeToSplit)
+
+  const UTC = timeAndUTC?.slice(timeToSplit + 7)
```
Comment (b-cooper) [inline app/src/organisms/ProtocolUpload/UploadInput.tsx]: We use [date-fns](https://date-fns.org/v2.25.0/docs) for manipulating date strings in our JS projects.  

By using a combination of their `parseIso` and `format` functions we should be able to get what we want in a cleaner, and less brittle manner.

```
import {format, parseIso} from 'date-fns'
const fullRunTimestamp = runQuery.data?.data.createdAt

dateFns.format(parseIso(fullRunTimestamp), 'Pppp')
```

File: app/src/assets/localization/en/protocol_info.json [hunk 2]
```diff
@@ -30,9 +30,21 @@
...
   "simulation_in_progress": "Simulation in Progress",
   "update_robot_for_custom_labware": "You have custom labware definitions saved to your app, but this robot needs to be updated before you can use these definitions with Python protocols",
-  "upload_and_simulate": "Upload and Simulate Protocol",
+  "upload_and_simulate": "Open a protocol to run on {{robot_name}}",
   "exit_modal_heading": "Confirm Close Protocol",
   "exit_modal_go_back": "No, go back",
   "exit_modal_exit": "Yes, close now",
-  "exit_modal_body": "Are you sure you want to close this protocol?"
+  "exit_modal_body": "Are you sure you want to close this protocol?",
+  "robotName_last_run": "{{robot_name}}’s last run",
```
Comment (shlokamin) [inline app/src/assets/localization/en/protocol_info.json]: nit: this key is a mix of snake case and camel case, we usually stick with snake case it looks like

File: app/src/assets/localization/en/protocol_info.json [hunk 3]
```diff
@@ -30,9 +30,21 @@
...
+  "protocol_name_title": "Protocol name",
+  "run_timestamp_title": "Run timestamp",
+  "labware_offset_data_title": "Labware Offset data",
+  "timestamp": "+{{index}}",
+  "run_again_btn": "Run again",
+  "labware_offsets_info": "{{number}} Labware Offsets",
+  "rerunning_protocol_modal_header": "How Rerunning A Protocol Works",
+  "rerunning_protocol_modal_body": "<block>Opentrons displays the connected robot’s last protocol run on on the Protocol Upload page. If you run again, Opentrons loads this protocol and applies Labware Offset data if any exists.</block><block>Clicking “Run Again” will take you directly to the Run tab. If you’d like to review the deck setup or run Labware Position Check before running the protocol, navigate to the Protocol tab.</block><block>If you recalibrate your robot, it will clear the last run from the upload page.</block>",
+  "rerunning_protocol_modal_link": "Learn more about Labware Offset Data",
+  "no_labware_offset_Data": "No Labware Offset data"
```
Comment (shlokamin) [inline app/src/assets/localization/en/protocol_info.json]: nit: i dont think `Data` should be capitalized in this key

File: app/src/organisms/ProtocolUpload/RerunningProtocolModal.tsx [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  DIRECTION_COLUMN,
+  SPACING_2,
+} from '@opentrons/components'
+import { Portal } from '../../App/portal'
+
+interface RerunningProtocolModalProps {
+  onCloseClick: () => unknown
+}
+
+const UPLOAD_PROTOCOL_URL = '#' //   TODO IMMEDIATELY get actual link - Emily said this url isn't created it
```
Comment (shlokamin) [inline app/src/organisms/ProtocolUpload/RerunningProtocolModal.tsx:33]: can we make a ticket for this so we don't forget?

File: app/src/organisms/ProtocolUpload/RerunningProtocolModal.tsx [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            {t('rerunning_protocol_modal_link')}
+            <Icon name={'open-in-new'} marginLeft={SPACING_1} size="10px" />
+          </Link>
+          <Box textAlign={ALIGN_CENTER} marginTop={SPACING_4}>
+            <PrimaryBtn
+              onClick={props.onCloseClick}
+              width={SIZE_4}
+              backgroundColor={C_BLUE}
+              name="close"
+              id={'RobotCalModal_closeButton'}
```
Comment (shlokamin) [inline app/src/organisms/ProtocolUpload/RerunningProtocolModal.tsx]: looks like this copied from `DeckCalibrationModal`, it should be prefixed with `RerunningProtocolModal`

File: app/src/organisms/ProtocolUpload/RerunningProtocolModal.tsx [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            components={{
+              block: <Text css={FONT_BODY_1_DARK} marginTop={SPACING_1} />,
+            }}
+          />
+          <Link
+            fontSize={FONT_SIZE_BODY_1}
+            color={C_BLUE}
+            href={UPLOAD_PROTOCOL_URL}
+            marginTop={SPACING_2}
+            id={'RerunningProtocol_ModalLink'}
```
Comment (shlokamin) [inline app/src/organisms/ProtocolUpload/RerunningProtocolModal.tsx]: this is should be prefixed with `RerunningProtocolModal`, right now theres an underscore between `Protocol` and `Modal`

File: app/src/organisms/ProtocolUpload/UploadInput.tsx [hunk 7]
```diff
@@ -56,12 +74,37 @@ export interface UploadInputProps {
...
+  const cloneMostRecentRun = useCloneRun(
+    mostRecentRunId != null ? mostRecentRunId : ''
+  )
+  //  If mostRecentRun is null, the CTA that uses cloneRun won't appear so this will never be reached
+  const robotName = useSelector((state: State) => getConnectedRobotName(state))
   const fileInput = React.useRef<HTMLInputElement>(null)
   const [isFileOverDropZone, setIsFileOverDropZone] = React.useState<boolean>(
     false
   )
+  const [rerunningProtocolModal, setRerunningProtocolModal] = React.useState(
```
Comment (shlokamin) [inline app/src/organisms/ProtocolUpload/UploadInput.tsx]: can these state variables be prefixed with `show` to indicate that they control whether or not the modal should be rendered?

File: app/src/organisms/ProtocolUpload/UploadInput.tsx [hunk 8]
```diff
@@ -56,12 +74,37 @@ export interface UploadInputProps {
...
+  const mostRecentRun = runQuery.data?.data
+  const mostRecentProtocolInfo = useProtocolQuery(
+    (mostRecentRun?.protocolId as string) ?? ''
+  )
+  const mostRecentProtocol = mostRecentProtocolInfo?.data?.data
+  const protocolData = useProtocolDetails()
+  const cloneMostRecentRun = useCloneRun(
+    mostRecentRunId != null ? mostRecentRunId : ''
+  )
+  //  If mostRecentRun is null, the CTA that uses cloneRun won't appear so this will never be reached
```
Comment (shlokamin) [inline app/src/organisms/ProtocolUpload/UploadInput.tsx]: nit: can this comment be moved to be just above line 87?

File: app/src/organisms/ProtocolUpload/index.tsx [hunk 9]
```diff
@@ -87,7 +89,14 @@ export function ProtocolUpload(): JSX.Element {
...
       }
     : {
-        title: t('upload_and_simulate'),
+        title: (
+          <Trans
+            id={'file_input'}
+            t={t}
+            i18nKey="upload_and_simulate"
+            values={{ robot_name: robotName }}
+          />
```
Comment (shlokamin) [inline app/src/organisms/ProtocolUpload/index.tsx]: how come this needs to be a `Trans` component?

**Review by shlokamin — APPROVED** [review]
left some nits, but no need to hold this up if its blocking other folks. also I did not test on a robot.

File: app/src/organisms/ProtocolUpload/index.tsx (hunk 9 above)
Comment (jerader) [reply app/src/organisms/ProtocolUpload/index.tsx]: can you add the `robot_name` without it being a `Trans` component?

Comment (shlokamin) [reply app/src/organisms/ProtocolUpload/index.tsx]: sure can! https://github.com/Opentrons/opentrons/blob/f0f31a96e3f7ce96c8b988c4bb108acd303214d6/app/src/organisms/ProtocolSetup/LabwarePositionCheck/hooks/useLabwarePositionCheck.ts#L68-L70

File: app/src/organisms/ProtocolUpload/RerunningProtocolModal.tsx (hunk 5 above)
Comment (jerader) [reply app/src/organisms/ProtocolUpload/RerunningProtocolModal.tsx]: whoops, good catch!

File: app/src/organisms/ProtocolUpload/index.tsx (hunk 9 above)
Comment (shlokamin) [reply app/src/organisms/ProtocolUpload/index.tsx]: Trans components are only really necessary when you need to interpolate React elements. most of the time they're actually unnecessary: 

`As long you have no React/HTML nodes integrated into a cohesive sentence (text formatting like strong, em, link components, maybe others), you won't need it - most of the times you will be using the good old t function.`

https://react.i18next.com/latest/trans-component

File: app/src/organisms/ProtocolUpload/RerunningProtocolModal.tsx (hunk 4 above)
Comment (jerader) [reply app/src/organisms/ProtocolUpload/RerunningProtocolModal.tsx:33]: just made one!

File: app/src/organisms/ProtocolUpload/index.tsx (hunk 9 above)
Comment (jerader) [reply app/src/organisms/ProtocolUpload/index.tsx]: nevermind, i see how to do it 😃

Comment (jerader) [reply app/src/organisms/ProtocolUpload/index.tsx]: whoops, replied to the comment without see your response. thanks for the info link!

**Pull request merged**